[Doc] APM doc first draft (#1618)
grafana/tempo: docs/tempo/website/metrics-generator/app-performance-mgmt.md

---
aliases:
- /docs/tempo/latest/server_side_metrics/service_graphs/
title: Application performance management
menuTitle: APM dashboard
weight: 200
---

# Application performance management dashboard

<span style="background-color:#f3f973;">This is an experimental feature. For more information about how to enable it, continue reading.</span>

Application performance management (APM) uses data gathered from services, agents, systems, and microservices to visualize and identify areas of concern.

The Grafana APM dashboard visualizes the span metrics (traces data for rates, error rates, and durations (RED)) and service graphs.
Once the requirements are set up, this pre-configured dashboard is immediately available.  

You can use the APM dashboard to:

* Discover spans which are consistently erroring and the rates at which they occur
* Get an overview of the overall rate of span calls throughout your services

> * Determine how long the top 10% slowest queries in your service take to complete

Reviewer: ```suggestion
* Determine how long the slowest queries in your service take to complete
```

I think in this case maybe it is just easier to be general? Makes the sentence a bit more readable and I figure should you choose you can adjust the percentile to show top X% where X is configurable.

Author: Thank you for the suggestion. Much appreciated.

* Examine all traces that contain spans of particular interest based on rate, error and duration values (RED signals)

<p align="center"><img src="../../getting-started/apm-overview.png" alt="APM dashboard"></p>

## Requirements to enable the APM dashboard

> You have to enable metric graph and service graph generation on the Grafana backend so metrics are generated as traces are ingested.

Author: ```suggestion
You have to enable span metrics and service graph generation on the Grafana backend so metrics are generated as traces are ingested.
```

To use the APM dashboard, you need:

> * Tempo or Grafana Cloud Traces with the metrics generator enabled and configured to send data to a Prometheus-compatible metrics store

Author: ```suggestion
* Tempo or Grafana Cloud Traces with the metrics generator or Grafana Agent enabled and configured to send data to a Prometheus-compatible metrics store
```

> * Tempo or Grafana Cloud Traces with the metrics generator enabled and configured to send data to a Prometheus compatible metrics

Reviewer: ```suggestion
* Tempo or Grafana Cloud Traces with the metrics generator enabled and configured to send data to a Prometheus-compatible metrics store.
```

* [Services graphs]({{< relref "../metrics-generator/service_graphs/#how-to-run" >}}), which are enabled by default in Grafana
* [APM table enabled](https://grafana.com/docs/grafana/next/datasources/tempo/) within the service graph tab (use the `tempoApmTable` feature flag in Grafana to enable)
* [Span metrics]({{< relref "../metrics-generator/span_metrics/#how-to-run" >}}) enabled in your Tempo data source configuration

> The APM dashboard can be generated with the metrics-generator or Grafana Agent.

Reviewer: If this is the case, doesn't this relax requirement (1)?

Namely 
```
Tempo or Grafana Cloud Traces with the metrics generator enabled and configured to send data to a Prometheus compatible metrics
```

Maybe requirement becomes "tempo or GCT with metrics generator...or grafana agent (with whatever transforms/extra stuff is required on the agent)"?

Author: This is one of the questions I have about this document. On Thursday, I learned that the APM dashboard will work with metrics-generator and/or Grafana Agent. I wrote this thinking just metrics-generator. It needs to be reviewed for that difference.

Author: I've made a suggestion above by adding the metrics-generator and Agent in the first requirements bullet.

Reviewer 2: I believe that the service graph and APM table and service graph will work with grafana agent metrics or tempo metrics, but it really should be tested.

Author: @joe-elliott APM table is marked as an experimental feature in this doc. Do you want the statement (first bullet) to stay as is or should it be reworded to only point to configuration that we have tested? And is that config metrics generator primarily?

Author: Clarified with Joe in Slack. Statement should be okay.

Any information in the APM table that has an underline can be selected to show more detailed information. You can also select any node in the service graph to display additional information. In the dashboard shown below, the `Ingester.QueryStream` span has a request rate of `144220.22` requests per second. The `cortex/IngesterQuery` span has the highest request rate.

### Error rate example

Let’s say we want to learn more about why `cortex.Ingester` has the highest error rates.
Selecting the second row of the Error rate column displays details about the span metrics in a new window on the right side.

<p align="center"><img src="../apm-error-rate-example.png" alt="APM error rate example"></p>

The metrics query used to generate the data appears in the **Metrics browser** field.

<p align="center"><img src="../apm-error-example-editor.png" alt="APM error example query editor"></p>

## Span metrics table

The span metrics, also called the APM table, are generated by the metrics-generator or the Grafana Agent. These metrics are created from ingested tracing data, including RED metrics.

Span metrics generate two metrics:

* A counter that computes requests

> * A histogram that tracks the distribution of durations of all requests per second

Reviewer: ```suggestion
* A histogram that tracks the distribution of durations of all requests
```

I think its not 'per second' is it? 
I think you just do distribution of request durations. The rate (requests per second) comes from looking at the change in the counter over time. 
But I may be totally wrong here :)

| Rate | LCD gauge (horizontal bar graph). Instances per second of the span. Clicking this field can jump to the appropriate metrics. |`sum(rate(  traces_spanmetrics_calls_total{ span_name="", <filters> }[$__range]))` |
| Error Rate | Number and LCD gauge (horizontal bar graph). Clicking this field shows more detailed metrics. | `sum(rate(  traces_spanmetrics_calls_total{ span_name="",   span_status="STATUS_CODE_ERROR", <filters> }[$__range]))` |
| Duration | p90 duration: 90% of all occurrences of this span complete within this time. Clicking this field shows the appropriate metrics. | `histogram_quantile(.9, sum(rate(  traces_spanmetrics_duration_seconds_bucket{ span_name="", span_status="STATUS_CODE_ERROR",  <filters> }[$__range]) by (le))` |
| Links | Provide links to example traces given the span name and other applied filters. Link to a search for all spans with the same name from the same Tempo data source. | N/A |

## Service graphs

A service graph (node graph) is a visual representation of the interrelationships between various services.
Service graphs help to understand the structure of a distributed system, and the connections and dependencies between its components.

> Service graphs infer the topology of a distributed system, provide a high level overview of the health of your system, and an historic view of a system’s topology.

Reviewer: ```suggestion
Service graphs infer the topology of a distributed system, provide a high level overview of the health of your system, and a historic view of a system’s topology.
```

Service graphs show error rates and latencies, among other relevant data.
The service graph layout can be the default or grid. 

<p align="center"><img src="../apm-service-graph-web.png" alt="APM service graph with a connected node layout"></p>

The grid layout changes the service graph to a series of rows and columns. 

<p align="center"><img src="../apm-service-graph-rows.png" alt="APM service graph with grid layout"></p>

If you are using the metrics-generator, then it processes traces and generates service graphs in the form of time series metrics like:

```yaml
tempo_service_graph_request_total{client="app", server="db"} 20
```

For information about service graphs and how they are calculated, refer to the [Service Graphs documentation]({{< relref "../metrics-generator/service_graphs.md" >}}).

## Use filters to reveal details

> The APM dashboard provides a gateway to your tracing information in the service graphs and span metrics. 

Reviewer: ```suggestion
The APM dashboard uses service graphs and span metrics to provide a gateway to your tracing information.
```

> his dashboard is derived from a fixed set of metrics queries.

Reviewer: ```suggestion
This dashboard is derived from a fixed set of metrics queries.
```

These underlying queries can not be changed.
However, you can choose which traces are included in the metrics query by filtering.

You can explore data by clicking on selectable items or by using filters.

### Selecting items or nodes for more detail

Clicking on selectable items, such as underlined text in the APM table or nodes on the service graph, lets you reveal specific details based upon your selection.

> In the APM table, you can select items in the **Rate**, **Error Rate**, **Duration (p90)**, and **Links** columns. Choosing one if these items provides details about the span metrics.

Reviewer: ```suggestion
In the APM table, you can select items in the **Rate**, **Error Rate**, **Duration (p90)**, and **Links** columns. Choosing one of these items provides details about the span metrics.
```

<p align="center"><img src="../apm-service-graph-drilldown.png" alt="APM service graph with drill-down"></p>

### Filter with metric queries

Using the filters at the top of the screen, you can narrow the data set based upon span attributes (key-value pairs or labels).
The filters build a query to refine what is shown in the service graph and span metrics.
You can add one or more label filters.

To use the filters:

1. At the top of the Service Graph, select the text box after **Filter** to display a list of available labels. In this case, **server** is selected. <br /><img src="../apm-query-filter-label.png" alt="APM service graph with grid layout">

1. Select or search for a value for the label. In this case, the value of **server** is equal to **tempo-ingester**. The default operator is equals (=). <br /> <img src="../apm-query-filter-value1.png" alt="APM select value for label">

1. Optional: Change the operator by selecting **=** and choosing a new option from the drop-down. <br /><img src="../apm-filter-operator.png" alt="APM filter operators">

> 1. Optional: Add additional key-value pairs to refine the data set. Any subsequent label filters use AND, which requires both key-value pairs to be presents for matches.

Reviewer: I guess that means "OR" is not supported?

Author: When you look at the screenshot, I did not find a way to change the AND statement to an OR.

Author: @joey-grafana Is there another way to change the AND?

Reviewer 3: It's just AND at them moment.

Author: Thank you.

1. Select **Run query**.

Filters can be removed by selecting the filter drop-down and choosing **– remove filter –**.

<p align="center"><img src="../apm-remove-filter.png" alt="APM remove filters"></p>

> In the example below, each field or label represents a key-value pair. Number 1 selects a service as the label whose value is Go-http-client (2). The second key-value pair has a client as a label whose value is 02e807.

Reviewer: ```suggestion
In the example below, each field or label represents a key-value pair. Number 1 selects a service as the label whose value is `Go-http-client` (2). The second key-value pair has a client as a label whose value is `02e807`.
```

Suggestion to make more readable. Idk if this conforms to our style guide though :)

Author: Nope, that works fine. And follows how I've been trying to do the other examples as well.

## Requirements to enable the APM dashboard

You have to enable span metrics and service graph generation on the Grafana backend so metrics that are generated as traces are ingested.

To use the APM dashboard, you need:

* Tempo or Grafana Cloud Traces with either 1) the metrics generator enabled and configured or 2) the Grafana Agent enabled and configured to send data to a Prometheus-compatible metrics store
* [Services graphs]({{< relref "../metrics-generator/service_graphs/#how-to-run" >}}), which are enabled by default in Grafana
* [APM table enabled](https://grafana.com/docs/grafana/next/datasources/tempo/) within the service graph tab (use the `tempoApmTable` feature flag in Grafana to enable)
* [Span metrics]({{< relref "../metrics-generator/span_metrics/#how-to-run" >}}) enabled in your Tempo data source configuration

The APM dashboard can be generated with the metrics-generator or Grafana Agent.

For information on how to configure these features, refer to the [Grafana Tempo data sources documentation](https://grafana.com/docs/grafana/next/datasources/tempo/).

## APM dashboard

> Using this dashboard, you can see the top five spans by request rate. You can refine any of this data using the filters. Selecting any of the data points lets you see more specific data.

Reviewer 2: "top five spans by request rate" - These should be filtered down to spans with a type of "server" only. The idea is that they represent endpoints instead of just any span.

@joey-grafana can you confirm the above?

Reviewer 3: I confirm that we are filtering down to only show spans with a type of server in the table.

> Using this dashboard, you can see the top five spans by request rate. You can refine any of this data using the filters. Selecting any of the data points lets you see more specific data.

Author: Is this correct? 
```suggestion
Using this dashboard, you can see the top five spans with a type of server (listed in the `Name` column).  
You can refine any of this data using the filters.
Selecting any of the data points lets you see more specific data.
```

The APM dashboard provides a span metrics visualization (APM table, screen section 2) and service graph (screen section 3). In addition, you can use the filters (screen section 1) to customize the data displayed.

<p align="center"><img src="../apm-overview-numbered.png" alt="APM dashboard with numbered sections"></p>

> Any information in the APM table that has an underline can be selected to show more detailed information. You can also select any node in the service graph to display additional information. In the dashboard shown below, the `Ingester.QueryStream` span has a request rate of `144220.22` requests per second. The `cortex/IngesterQuery` span has the highest request rate.

Reviewer 2: cortex -> mimir?

Author: Good catch. I hadn't realized the text didn't match the screen shot. Fixed that.

> Any information in the APM table that has an underline can be selected to show more detailed information. You can also select any node in the service graph to display additional information. In the dashboard shown below, the `Ingester.QueryStream` span has a request rate of `144220.22` requests per second. The `cortex/IngesterQuery` span has the highest request rate.

Author: ```suggestion
Any information in the APM table that has an underline can be selected to show more detailed information.
You can also select any node in the service graph to display additional information.
In the dashboard shown below, the `Ingester.QueryStream` span has a request rate of `144220.22` requests per second.
The `/cortex.Ingester/Query` span has the highest request rate.
```
